File: topoflow/trace.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Discover Legal
"""Two-level trace & RunReport (spec §11). JSON-serializable throughout."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from .types import Action, Signature


@dataclass
class Trace:
    task_id: str
    events: list[dict] = field(default_factory=list)
    tokens: int = 0
    retries: int = 0
    any_failure: bool = False
    final_answer: str = ""
    # macro decision path -> list[(Signature, Action)] for backup
    _decisions: list[tuple] = field(default_factory=list)

    def add(self, event: dict) -> None:
        self.events.append(event)

    def record_decision(self, sig: Signature, action: Action) -> None:
        self._decisions.append((sig, action))

    def decision_path(self) -> list[tuple]:
        return list(self._decisions)

    def to_dict(self) -> dict:
        return {
            "task_id": self.task_id,
            "tokens": self.tokens,
            "retries": self.retries,
            "any_failure": self.any_failure,
            "final_answer": self.final_answer,
            "events": self.events,
        }


# ── event builders ──────────────────────────────────────────────────────────
def invoke_event(action: Action, out) -> dict:
    return {
        "type": "invoke",
        "skill": action.skill,
        "model": action.model,
        "tokens": out.tokens,
        "complete": out.complete,
        "failed": out.failed,
        "delta_fields": sorted(out.handoff_delta.keys()),
    }


def skip_event(action: Action) -> dict:
    return {"type": "skip", "target": action.target}


def topology_event(action: Action, res) -> dict:
    return {
        "type": "topology",
        "mode": action.topo_mode,
        "tau_bucket": action.tau_bucket,
        "k_in": action.k_in,
        "round_bucket": action.round_bucket,
        "rounds_run": res.rounds_run,
        "total_tokens": res.total_tokens,
        "retries": res.retries,
        "failed": res.failed,
        "subtrace": res.subtrace,
    }


@dataclass
class RunReport:
    task_id: str
    reward: float
    quality: float
    sub_scores: dict
    tokens: int
    retries: int
    decision_path: list[dict]
    aborted: bool = False
    abort_reason: Optional[str] = None
    trace: Optional[dict] = None

    def to_dict(self) -> dict:
        return {
            "task_id": self.task_id,
            "reward": self.reward,
            "quality": self.quality,
            "sub_scores": self.sub_scores,
            "tokens": self.tokens,
            "retries": self.retries,
            "decision_path": self.decision_path,
            "aborted": self.aborted,
            "abort_reason": self.abort_reason,
            "trace": self.trace,
        }


def build_run_report(ctx_task_id, trace: Trace, reward, quality, sub, aborted=False, reason=None) -> RunReport:
    dp = [
        {"signature": _sig_brief(s), "action": list(a.key())} for (s, a) in trace.decision_path()
    ]
    return RunReport(
        task_id=ctx_task_id,
        reward=reward,
        quality=quality,
        sub_scores=sub,
        tokens=trace.tokens,
        retries=trace.retries,
        decision_path=dp,
        aborted=aborted,
        abort_reason=reason,
        trace=trace.to_dict(),
    )


def _sig_brief(sig: Signature) -> str:
    return f"{sig.regime.value}|m{''.join(map(str, sig.handoff_mask))}"

File: biglaw-go/internal/topoflow/trace.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 Discover Legal

package topoflow

// decision is one macro decision-path entry (for backup).
type decision struct {
	Sig Signature
	Act Action
}

// Trace is the two-level, JSON-serializable record (spec §11).
type Trace struct {
	TaskID      string           `json:"taskId"`
	Events      []map[string]any `json:"events"`
	Tokens      int              `json:"tokens"`
	Retries     int              `json:"retries"`
	AnyFailure  bool             `json:"anyFailure"`
	FinalAnswer string           `json:"finalAnswer"`
	decisions   []decision
}

// NewTrace creates an empty trace.
func NewTrace(taskID string) *Trace { return &Trace{TaskID: taskID} }

func (t *Trace) add(ev map[string]any) { t.Events = append(t.Events, ev) }
func (t *Trace) recordDecision(s Signature, a Action) {
	t.decisions = append(t.decisions, decision{s, a})
}

// DecisionPath returns the macro decisions in order (for backup).
func (t *Trace) DecisionPath() []decision { return t.decisions }

func invokeEvent(a Action, out CellOutput) map[string]any {
	keys := make([]string, 0, len(out.HandoffDelta))
	for k := range out.HandoffDelta {
		keys = append(keys, k)
	}
	return map[string]any{
		"type": "invoke", "skill": a.Skill, "model": a.Model,
		"tokens": out.Tokens, "complete": out.Complete, "failed": out.Failed,
		"deltaFields": keys,
	}
}

func skipEvent(a Action) map[string]any {
	return map[string]any{"type": "skip", "target": a.Target}
}

func topologyEvent(a Action, res TopologyResult) map[string]any {
	return map[string]any{
		"type": "topology", "mode": a.TopoMode,
		"tauBucket": a.TauBucket, "kIn": a.KIn, "roundBucket": a.RoundBucket,
		"rounds_run": res.RoundsRun, "total_tokens": res.TotalTokens,
		"retries": res.Retries, "failed": res.Failed, "subtrace": res.Subtrace,
	}
}

// RunReport is the per-task artifact (spec §11), emitted even on governance abort.
type RunReport struct {
	TaskID       string           `json:"taskId"`
	Reward       float64          `json:"reward"`
	Quality      float64          `json:"quality"`
	SubScores    map[string]any   `json:"subScores"`
	Tokens       int              `json:"tokens"`
	Retries      int              `json:"retries"`
	DecisionPath []map[string]any `json:"decisionPath"`
	Aborted      bool             `json:"aborted"`
	AbortReason  string           `json:"abortReason,omitempty"`
	Trace        *Trace           `json:"trace"`
}

func buildRunReport(t *Trace, reward, quality float64, sub map[string]any, aborted bool, reason string) *RunReport {
	dp := make([]map[string]any, 0, len(t.decisions))
	for _, d := range t.decisions {
		dp = append(dp, map[string]any{"signature": sigLabel(d.Sig), "action": d.Act})
	}
	return &RunReport{
		TaskID: t.TaskID, Reward: reward, Quality: quality, SubScores: sub,
		Tokens: t.Tokens, Retries: t.Retries, DecisionPath: dp,
		Aborted: aborted, AbortReason: reason, Trace: t,
	}
}
